cluster/v1alpha1/helpers.go

```diff
@@ -55,24 +54,30 @@ type ClusterRolloutStatus struct {
 
 // RolloutResult contains the clusters to be rolled out and the clusters that have timed out.
 type RolloutResult struct {
-	// ClustersToRollout is a map where the key is the cluster name and the value is the ClusterRolloutStatus.
-	ClustersToRollout map[string]ClusterRolloutStatus
-	// ClustersTimeOut is a map where the key is the cluster name and the value is the ClusterRolloutStatus.
-	ClustersTimeOut map[string]ClusterRolloutStatus
+	// ClustersToRollout is a slice of ClusterRolloutStatus that will be rolled out.
+	ClustersToRollout []ClusterRolloutStatus
+	// ClustersTimeOut is a slice of ClusterRolloutStatus that are timeout.
+	ClustersTimeOut []ClusterRolloutStatus
+	// ClustersRemoved is a slice of ClusterRolloutStatus that are removed.
+	ClustersRemoved []ClusterRolloutStatus
 }
 
+// ClusterRolloutStatusFunc defines a function to return the rollout status for a given workload.
+type ClusterRolloutStatusFunc[T any] func(clusterName string, workload T) (ClusterRolloutStatus, error)
+
 // +k8s:deepcopy-gen=false
-type RolloutHandler struct {
+type RolloutHandler[T any] struct {
```

Comment:
but any is a more loose constraint, isn't it?

Reply:
Well, yes "any" is kind of loose, but we are safe. The main idea behind make a generic Type "any" is to avoid doing casting (or force certain type) at the clusterRolloutStatusFunc Function implementation similar to the unit test example [here](https://github.com/serngawy/api/blob/rolloutLib/cluster/v1alpha1/helpers_test.go#L47) AND the helper lib is safe we don't call the clusterRolloutStatusFunc any where, its just a func definition that will be used to create the existing ClustersRollOutStatus at the consumer API side. What do you think ? Is that make sense ?

Reply:
@xuezhaojun I noticed that also and ran `controller-gen` to generate the functions instead--would that work instead here (at least until `gengo` supports generics or even migrate the whole thing to use `controller-gen`)?
```shell
./_output/tools/bin/controller-gen object:headerFile="hack/empty.txt" paths="./cluster/v1alpha1"
```

Reply:
Hi, @dhaiducek , I test `controller-gen` with latest api main branch, it fails and generates unexpected code for generic type:
```
./_output/tools/bin/controller-gen object:headerFile="hack/empty.txt" paths="./cluster/v1alpha1"
open-cluster-management.io/api/cluster/v1alpha1:-: invalid type: func(clusterName string, workload T) (open-cluster-management.io/api/cluster/v1alpha1.ClusterRolloutStatus, error)
Error: not all generators ran successfully
run `controller-gen object:headerFile=hack/empty.txt paths=./cluster/v1alpha1 -w` to see all available markers, or `controller-gen object:headerFile=hack/empty.txt paths=./cluster/v1alpha1 -h` for usage
```

Reply:
As I mentioned above, we can also change to the explicit way to claim which struct needs gen deepcopy funcs: https://github.com/open-cluster-management-io/api/pull/288

Reply:
Actually it seems the `gengo` tooling may not be respecting the `deepcopy-gen=false` tag? In any case, I don't have a great deal of familiarity with deepcopy, but here's a commit that might work as an alternative to selectively generating the APIs:
- https://github.com/dhaiducek/api/commit/3734d0e19b78b8feb430640486103e520a15a927

Reply:
Thanks @xuezhaojun I'm able to reproduce it as well. having the same issue as @dhaiducek sounds like deepCopy generator does not respect the `deepcopy-gen=false` tag ! 
Based on the deepCopy code doc [here](https://github.com/kubernetes/code-generator/blob/master/cmd/deepcopy-gen/main.go#L39) setting the deepcopy-gen=false should work.

Reply:
Thanks for providing the alternative! I will follow with it.